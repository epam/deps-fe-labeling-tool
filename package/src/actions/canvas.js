import { createAction } from 'redux-actions'

const FEATURE_NAME = 'CANVAS'

const setZoom = createAction(
  `${FEATURE_NAME}/SET_ZOOM`
)

const setRotationAngle = createAction(
  `${FEATURE_NAME}/SET_ROTATION_ANGLE`,
  (page, angle) => ({
    page,
    angle
  })
)

const setScale = createAction(
  `${FEATURE_NAME}/SET_SCALE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  setZoom,
  setScale,
  setRotationAngle
}
