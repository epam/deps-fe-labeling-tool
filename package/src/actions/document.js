import { createAction } from 'redux-actions'

const FEATURE_NAME = 'DOCUMENT'

const storeDocument = createAction(`${FEATURE_NAME}/STORE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  storeDocument
}
