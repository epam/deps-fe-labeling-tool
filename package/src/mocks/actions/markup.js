const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
