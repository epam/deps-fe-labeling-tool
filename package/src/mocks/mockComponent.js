const mockComponent = (componentName, namedExport = true, fn = () => { }) => {
  Object.defineProperty(fn, 'name', { value: componentName })
  return !namedExport ? fn : { [componentName]: fn }
}

export {
  mockComponent
}
