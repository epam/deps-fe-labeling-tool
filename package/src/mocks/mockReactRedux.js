const mockDispatch = jest.fn((action) => action)

const createMockMapStateToProps = (mapStateToProps) => {
  if (!mapStateToProps) {
    return mapStateToProps
  }

  return (state, ownProps) => ({
    props: mapStateToProps(state, ownProps)
  })
}

const getDispatchFunctionFormShorthandObject = (actionsObj) => (dispatch) => (
  Object.keys(actionsObj).reduce(
    (props, actionKey) => ({
      ...props,
      [actionKey]: jest.fn(
        (...args) => dispatch(actionsObj[actionKey].apply(undefined, args))
      )
    }),
    {}
  )
)

const createMockMapDispatchToProps = (mapDispatchToProps) => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps
  }

  const mappingFunction = typeof mapDispatchToProps === 'function'
    ? mapDispatchToProps
    : getDispatchFunctionFormShorthandObject(mapDispatchToProps)

  return (dispatch, ownProps) => {
    dispatch = dispatch || mockDispatch

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    }
  }
}

const createMockMergeProps = (mergeProps) => {
  if (!mergeProps) {
    return mergeProps
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: mockDispatch }

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    }
  }
}

const mockReactRedux = {
  connect: (mapStateToProps, mapDispatchToProps, mergeProps) => (WrappedComponent) => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    WrappedComponent
  }),
  batch: (fn) => Promise.resolve(fn),
  useDispatch: jest.fn(() => mockDispatch),
  useSelector: jest.fn((selector) => selector())
}

export {
  mockReactRedux
}
