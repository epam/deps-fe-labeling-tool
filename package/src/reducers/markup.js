import has from 'lodash/has'
import { handleActions } from 'redux-actions'
import {
  insertCopiedMarkup,
  resetDefault,
  addLabels,
  storeMarkup,
  storeAssignedMarkup,
  storeImportMarkup,
  updateLabels,
  removeLabels,
  selectLabels,
  clearSelection,
  addTables,
  selectTables,
  updateTables,
  removeTables,
  updateAllLabels,
  updateAllTables,
  updateInitialMarkup
} from '@/actions/markup'
import { Label, LabelType } from '@/models/Label'
import { Table } from '@/models/Table'
import { isEqual } from '@/utils/isEqual'

const defaultState = {}

const setModifiedObjectsDuringUpdates = (updatedObjects, initialObjects, modifiedObjects) => {
  let newModifiedObjects = [...modifiedObjects]
  updatedObjects.forEach((object) => {
    const initialObject = initialObjects.find((io) => io.uid === object.uid)
    if (isEqual(object, initialObject)) {
      newModifiedObjects = newModifiedObjects.filter((mo) => mo !== object.uid)
    } else {
      newModifiedObjects = [...new Set([...newModifiedObjects, object.uid])]
    }
  })
  return newModifiedObjects
}

const setModifiedObjectsDuringRemove = (idsToRemove, initialObjects, modifiedObjects) => {
  let newModifiedObjects = [...modifiedObjects]
  idsToRemove.forEach((id) => {
    const initialObject = initialObjects.find((initialObject) => initialObject.uid === id)
    if (initialObject && !newModifiedObjects.includes(initialObject.uid)) {
      newModifiedObjects.push(id)
    } else {
      newModifiedObjects = newModifiedObjects.filter((mo) => mo !== id)
    }
  })
  return newModifiedObjects
}

const addLabelsHandler = (state, action) => {
  const modifiedObjects = [...state.modifiedObjects]

  action.payload.labels.forEach((label) => {
    modifiedObjects.push(label.uid)
  })

  return {
    ...state,
    modifiedObjects,
    labels: [
      ...state.labels,
      ...action.payload.labels
    ]
  }
}

const storeMarkupHandler = (state, action) => {
  const markup = action.payload
  return Object.entries(markup).reduce((newState, [page, pageMarkup]) => ({
    ...newState,
    [page]: {
      ...state[page],
      labels: (pageMarkup && pageMarkup.labels) || [],
      initialLabels: (pageMarkup && pageMarkup.labels) || [],
      tables: (pageMarkup && pageMarkup.tables) || [],
      initialTables: (pageMarkup && pageMarkup.tables) || [],
      modifiedObjects: [],
      selectedLabelsIds: [],
      selectedTablesIds: []
    }
  }), {})
}

const updateMarkupHandler = (state, action) => {
  const getUnassignedMarkup = (markup) => markup.filter((m) => m.fieldCode === '')

  const getUpdatedMarkup = (oldMarkup, newMarkup) => ({
    labels: [
      ...newMarkup.labels || [],
      ...getUnassignedMarkup(oldMarkup.labels)
    ],
    tables: [
      ...newMarkup.tables || [],
      ...getUnassignedMarkup(oldMarkup.tables)
    ]
  })

  return Object.entries(action.payload).reduce((newState, [page, pageMarkup]) => {
    const updatedMarkup = getUpdatedMarkup(state[page], pageMarkup)

    return {
      ...newState,
      [page]: {
        ...state[page],
        labels: updatedMarkup.labels,
        initialLabels: updatedMarkup.labels,
        tables: updatedMarkup.tables,
        initialTables: updatedMarkup.tables,
        modifiedObjects: [],
        selectedLabelsIds: [],
        selectedTablesIds: []
      }
    }
  }, {})
}

const storeImportMarkupHandler = (state, action) => {
  const markup = action.payload
  return Object.entries(markup).reduce((newState, [page, pageMarkup]) => ({
    ...newState,
    [page]: (() => {
      let modifiedObjects = []
      if (pageMarkup) {
        modifiedObjects = pageMarkup.labels ? [...modifiedObjects, ...pageMarkup.labels.map((label) => label.uid)] : modifiedObjects
        modifiedObjects = pageMarkup.tables ? [...modifiedObjects, ...pageMarkup.tables.map((table) => table.uid)] : modifiedObjects
      }
      return {
        ...state[page],
        labels: (pageMarkup && pageMarkup.labels) || [],
        initialLabels: [],
        tables: (pageMarkup && pageMarkup.tables) || [],
        initialTables: [],
        modifiedObjects,
        selectedLabelsIds: [],
        selectedTablesIds: []
      }
    })()
  }), {})
}

const getUpdatedStateByCriteria = (markup, criteria, props) => {
  const criteriaEntries = Object.entries(criteria)
  const updatedObjects = {}

  const update = (objects, page, initialKey) => {
    let updatedObject

    return objects.map((o) => {
      if (criteriaEntries.every(([key, value]) => has(o, key) && o[key] === value)) {
        updatedObject = {
          ...o,
          ...props
        }
        let modifiedObjects = [...markup[page].modifiedObjects]
        const initialObject = markup[page][initialKey].find((initial) => initial?.uid === o.uid) ?? {}
        if (isEqual(initialObject, updatedObject) && modifiedObjects.includes(o.uid)) {
          modifiedObjects = modifiedObjects.filter((mo) => mo !== o.uid)
        } else {
          modifiedObjects = [...new Set([...modifiedObjects, o.uid])]
        }
        updatedObjects[page] = modifiedObjects
        return updatedObject
      }
      return o
    })
  }

  const updatedState = Object.entries(markup).reduce((newState, [page, pageMarkup]) => ({
    ...newState,
    [page]: {
      ...markup[page],
      labels: update(pageMarkup.labels, page, 'initialLabels'),
      tables: update(pageMarkup.tables, page, 'initialTables')
    }
  }), {})

  if (!isEqual(updatedObjects, {})) {
    for (const [key, value] of Object.entries(updatedObjects)) {
      updatedState[key].modifiedObjects = value
    }
  }

  return updatedState
}

const updateLabelsHandler = (state, action) => {
  const modifiedObjects = setModifiedObjectsDuringUpdates(
    action.payload.labels,
    state.initialLabels,
    state.modifiedObjects
  )

  const updatedLabels = state.labels.map((currentLabel) => {
    const updatedLabel = action.payload.labels.find((l) => l.uid === currentLabel.uid)
    if (!updatedLabel) {
      return currentLabel
    }

    return {
      ...currentLabel,
      ...updatedLabel
    }
  })

  return ({
    ...state,
    modifiedObjects,
    labels: updatedLabels
  })
}
const removeLabelsHandler = (state, action) => {
  const idsToRemove = action.payload.labels.map((label) => label.uid)
  const modifiedObjects = setModifiedObjectsDuringRemove(
    idsToRemove,
    state.initialLabels,
    state.modifiedObjects
  )

  return {
    ...state,
    modifiedObjects,
    labels: state.labels.filter((label) => (
      !idsToRemove.includes(label.uid)
    )),
    selectedLabelsIds: state.selectedLabelsIds.filter((uid) => (
      !idsToRemove.includes(uid)
    ))
  }
}

const selectLabelsHandler = (state, action) => ({
  ...state,
  selectedLabelsIds: action.payload.labels.map((obj) => obj.uid)
})

const addTablesHandler = (state, action) => {
  const modifiedObjects = [...state.modifiedObjects]

  action.payload.tables.forEach((table) => {
    modifiedObjects.push(table.uid)
  })

  return ({
    ...state,
    modifiedObjects,
    tables: [
      ...state.tables,
      ...action.payload.tables
    ]
  })
}

const selectTablesHandler = (state, action) => ({
  ...state,
  selectedTablesIds: action.payload.tables.map((table) => table.uid)
})

const updateTablesHandler = (state, action) => {
  const modifiedObjects = setModifiedObjectsDuringUpdates(
    action.payload.tables,
    state.initialTables,
    state.modifiedObjects
  )

  return ({
    ...state,
    modifiedObjects,
    tables: state.tables.map((currentTable) => {
      const updatedTable = action.payload.tables.find((table) => table.uid === currentTable.uid)
      if (!updatedTable) {
        return currentTable
      }
      return {
        ...updatedTable
      }
    })
  })
}

const removeTablesHandler = (state, action) => {
  const idsToRemove = action.payload.tables.map((table) => table.uid)
  const modifiedObjects = setModifiedObjectsDuringRemove(
    idsToRemove,
    state.initialTables,
    state.modifiedObjects
  )

  return {
    ...state,
    modifiedObjects,
    tables: state.tables.filter((table) => (
      !idsToRemove.includes(table.uid)
    )),
    selectedTablesIds: state.selectedTablesIds.filter((uid) => (
      !idsToRemove.includes(uid)
    ))
  }
}

const clearSelectionHandler = (state) => ({
  ...state,
  selectedLabelsIds: [],
  selectedTablesIds: []
})

const resetDefaultHandler = () => (defaultState)

const updateInitialMarkupHandler = (state, action) => {
  return Object.entries(state).reduce((newState, [page, pageMarkup]) => ({
    ...newState,
    [page]: {
      ...state[page],
      initialLabels: (pageMarkup && pageMarkup.labels) || [],
      initialTables: (pageMarkup && pageMarkup.tables) || [],
      modifiedObjects: []
    }
  }
  ), {})
}

const resetObjectAndGetNewState = (object, state) => {
  const criteria = {
    uid: object.uid
  }

  const update = {
    fieldCode: '',
    type: LabelType.UNASSIGNED,
    index: undefined
  }

  return getUpdatedStateByCriteria(state, criteria, update)
}

const updateObjectAndGetNewState = (object, state, fieldCode, index, type, content) => {
  const criteria = {
    uid: object.uid
  }

  const update = {
    fieldCode,
    index
  }

  content !== undefined && (
    update.content = content
  )

  type && (update.type = type)

  return getUpdatedStateByCriteria(state, criteria, update)
}

const wrapInPayload = (payload) => ({ payload })

// TODO: #2700
const updateAllLabelsHandler = (state, action) => {
  const { multiAssign, label } = action.payload
  let newState = state

  if (!multiAssign) {
    const allLabels = Object.entries(newState).reduce(
      (acc, [page, pageMarkup]) => ([...acc, ...pageMarkup.labels]),
      []
    )

    const sameLabel = Label.getSameLabel(label, allLabels)
    newState = sameLabel ? resetObjectAndGetNewState(sameLabel, newState) : newState
  }

  return updateObjectAndGetNewState(
    label,
    newState,
    label.fieldCode,
    label.index,
    label.type,
    label.content
  )
}

// TODO: #2700
const updateAllTablesHandler = (state, action) => {
  const { multiAssign, table } = action.payload
  let newState = state

  if (!multiAssign) {
    const allTables = Object.entries(newState).reduce(
      (acc, [page, pageMarkup]) => ([...acc, ...pageMarkup.tables]),
      []
    )

    const sameTable = Table.getSameTable(table, allTables)
    newState = sameTable ? resetObjectAndGetNewState(sameTable, newState) : newState
  }

  return updateObjectAndGetNewState(
    table,
    newState,
    table.fieldCode,
    table.index
  )
}

const insertCopiedMarkupHandler = (state, action) => {
  const { page, pageMarkup } = action.payload

  let newState = state

  if (!newState[page]) {
    newState = {
      ...newState,
      [page]: defaultPageState
    }
  }

  if (pageMarkup.labels?.length) {
    const labelsToInsert = pageMarkup.labels.map((label) => {
      const duplicatedLabel = Label.duplicate(label)
      return Label.shift(duplicatedLabel)
    })

    newState = {
      ...newState,
      [page]: addLabelsHandler(
        newState[page],
        wrapInPayload({ labels: labelsToInsert })
      )
    }
    newState = {
      ...newState,
      [page]: selectLabelsHandler(
        newState[page],
        wrapInPayload({ labels: labelsToInsert })
      )
    }
  }

  if (pageMarkup.tables?.length) {
    const tablesToInsert = pageMarkup.tables.map((table) => {
      const duplicatedTable = Table.duplicate(table)
      const siftedTable = Table.shift(duplicatedTable)
      return siftedTable
    })

    newState = {
      ...newState,
      [page]: addTablesHandler(
        newState[page],
        wrapInPayload({ tables: tablesToInsert })
      )
    }
    newState = {
      ...newState,
      [page]: selectTablesHandler(
        newState[page],
        wrapInPayload({ tables: tablesToInsert })
      )
    }
  }
  return newState
}

// TODO: #1431
const pageMarkupReducerMap = new Map([
  [
    addLabels,
    addLabelsHandler
  ],
  [
    insertCopiedMarkup,
    insertCopiedMarkupHandler
  ],
  [
    resetDefault,
    resetDefaultHandler
  ],
  [
    storeMarkup,
    storeMarkupHandler
  ],
  [
    storeImportMarkup,
    storeImportMarkupHandler
  ],
  [
    updateLabels,
    updateLabelsHandler
  ],
  [
    removeLabels,
    removeLabelsHandler
  ],
  [
    selectLabels,
    selectLabelsHandler
  ],
  [
    clearSelection,
    clearSelectionHandler
  ],
  [
    addTables,
    addTablesHandler
  ],
  [
    selectTables,
    selectTablesHandler
  ],
  [
    removeTables,
    removeTablesHandler
  ],
  [
    updateTables,
    updateTablesHandler
  ],
  [
    updateInitialMarkup,
    updateInitialMarkupHandler
  ],
  [
    updateAllTables,
    updateAllTablesHandler
  ],
  [
    updateAllLabels,
    updateAllLabelsHandler
  ],
  [
    storeAssignedMarkup,
    updateMarkupHandler
  ]
])

const defaultPageState = {
  labels: [],
  tables: [],
  initialLabels: [],
  initialTables: [],
  selectedLabelsIds: [],
  selectedTablesIds: [],
  modifiedObjects: []
}

const pageMarkupReducer = handleActions(
  pageMarkupReducerMap,
  defaultPageState
)

const storeMarkupReducer = handleActions(
  pageMarkupReducerMap,
  defaultState
)

const markupReducer = (state, action) => {
  if (!state) {
    return defaultState
  }

  const knownActions = Array.from(pageMarkupReducerMap.keys(), (k) => k.toString())
  if (!knownActions.includes(action && action.type)) {
    return state
  }

  if (
    action.type === insertCopiedMarkup.toString() ||
    action.type === storeMarkup.toString() ||
    action.type === storeAssignedMarkup.toString() ||
    action.type === storeImportMarkup.toString() ||
    action.type === resetDefault.toString() ||
    action.type === updateAllLabels.toString() ||
    action.type === updateAllTables.toString() ||
    action.type === updateInitialMarkup.toString()
  ) {
    return storeMarkupReducer(state, action)
  }

  const page = (action.payload && action.payload.page) || action.payload
  if (!Number.isInteger(page)) {
    return state
  }

  return {
    ...state,
    [page]: pageMarkupReducer(state[page] || defaultPageState, action)
  }
}

export {
  markupReducer
}
