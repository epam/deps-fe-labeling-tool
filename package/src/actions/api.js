import { createAction } from 'redux-actions'
import { rotationAnglesSelector } from '@/selectors/canvas'
import { pageImageUrlSelector } from '@/selectors/document'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { primaryLanguageSelector } from '@/selectors/ocr'
import { getApi } from '@/services/api'

const FEATURE_NAME = 'API'

const requestAttempt = createAction(
  `${FEATURE_NAME}/REQUEST_ATTEMPT`
)

const requestSuccess = createAction(
  `${FEATURE_NAME}/REQUEST_SUCCESS`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const requestFailure = createAction(
  `${FEATURE_NAME}/REQUEST_FAILURE`,
  (requestId, error) => ({
    requestId,
    error
  })
)

export const createRequestAction = (requestId, actionCreator) => {
  const requestActionCreator = (...args) => async (dispatch, getState) => {
    try {
      dispatch(requestAttempt(requestId))
      const result = await actionCreator(...args)(dispatch, getState)
      dispatch(requestSuccess(requestId))
      return result
    } catch (error) {
      dispatch(requestFailure(requestId, error.message))
      throw error
    }
  }
  requestActionCreator.toString = () => requestId
  return requestActionCreator
}

const save = createRequestAction(
  'save',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().save(markup, rotationAngles, language)
  }
)

const saveMarkup = createRequestAction(
  'saveMarkup',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    const fields = fieldsSelector(state)
    return getApi().saveMarkup(markup, rotationAngles, language, fields)
  }
)

const ocrText = createRequestAction(
  'ocrText',
  (engine, coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrText(
      engine,
      pageImageUrl,
      coordinates,
      language
    )
  }
)

const omrArea = createRequestAction(
  'omrArea',
  (coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)

    return getApi().omrArea(
      pageImageUrl,
      coordinates
    )
  }
)

const detectTables = createRequestAction(
  'detectTables',
  (selectedAreaCoords) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    return getApi().detectTables(
      pageImageUrl,
      selectedAreaCoords
    )
  }
)

const ocrTable = createRequestAction(
  'ocrTable',
  (engine, table) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrTable(
      engine,
      pageImageUrl,
      table,
      language
    )
  }
)

export {
  detectTables,
  ocrTable,
  ocrText,
  omrArea,
  resetDefault,
  requestAttempt,
  requestSuccess,
  requestFailure,
  save,
  saveMarkup
}
