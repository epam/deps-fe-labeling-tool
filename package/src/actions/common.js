import { resetDefault as resetApi } from './api'
import { resetDefault as resetCanvas } from './canvas'
import { resetDefault as resetDocument } from './document'
import { resetDefault as resetImage } from './image'
import { resetDefault as resetMarkup } from './markup'
import { resetDefault as resetModel } from './model'
import { resetDefault as resetOcr } from './ocr'
import { resetDefault as resetPagination } from './pagination'
import { resetDefault as resetSettings } from './settings'
import { resetDefault as resetTools } from './tools'

const resetDefault = () => (dispatch) => {
  dispatch(resetApi())
  dispatch(resetOcr())
  dispatch(resetCanvas())
  dispatch(resetImage())
  dispatch(resetDocument())
  dispatch(resetMarkup())
  dispatch(resetPagination())
  dispatch(resetTools())
  dispatch(resetModel())
  dispatch(resetSettings())
}

export {
  resetDefault
}
