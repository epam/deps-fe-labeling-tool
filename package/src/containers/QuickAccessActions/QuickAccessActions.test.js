import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import React from 'react'
import { shallow } from 'enzyme'
import { undo } from '@/actions/markup'
import { QuickAccessActions } from '.'

const { WrappedComponent, mapDispatchToProps } = QuickAccessActions

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/containers/SaveDropdown', () => mockComponent('SaveDropdown'))

describe('Container: QuickAcessActions', () => {
  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass undo action as prop to WrappedComponent', () => {
      props.undo()
      expect(undo).toHaveBeenCalledTimes(1)
    })
  })

  describe('component', () => {
    let wrapper, defaultProps
    beforeEach(() => {
      defaultProps = {
        ...mapDispatchToProps().props
      }
      delete defaultProps.modifiedObjects

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render layout correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
