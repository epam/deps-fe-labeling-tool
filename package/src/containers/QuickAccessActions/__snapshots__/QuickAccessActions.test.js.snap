// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: QuickAcessActions component should render layout correctly 1`] = `
<Fragment>
  <SaveDropdown />
  <CompilationFeatureControl
    featureName="SHOW_NOT_IMPLEMENTED"
  >
    <Tooltip
      title="View"
    >
      <IconButton
        icon={<ViewIcon />}
      />
    </Tooltip>
    <Tooltip
      title="Apply template"
    >
      <IconButton
        icon={<TemplateIcon />}
      />
    </Tooltip>
  </CompilationFeatureControl>
  <Tooltip
    title="Revert to last saved version (Ctrl+Z)"
  >
    <IconButton
      icon={<RevertIcon />}
      onClick={[MockFunction]}
    />
  </Tooltip>
</Fragment>
`;
